app/scripts/migrations/165.ts

```diff
@@ -0,0 +1,309 @@
+
+  versionedData.data = transformState(versionedData.data);
+
+  return versionedData;
+}
+
+type LegacyStoredJobInformation = {
+  lastRun: number;
+};
+
+type LegacyBackgroundEvent = {
+  id: string;
+  scheduledAt: string;
+  snapId: SnapId;
+  date: string;
+  request: {
+    method: string;
+    jsonrpc?: '2.0' | undefined;
+    id?: string | number | null | undefined;
+    params?: Json[] | Record<string, Json> | undefined;
+  };
+};
+
+type LegacyCronjobControllerState = {
+  // `jobs` is optional to support `delete` operations in the migration.
+  jobs?: Record<string, LegacyStoredJobInformation>;
+  events: Record<string, LegacyBackgroundEvent>;
+};
+
+/**
+ * Represents a background event that is scheduled to be executed by the
+ * cronjob controller.
+ */
+type BackgroundEvent = LegacyBackgroundEvent & {
+  /**
+   * Whether the event is recurring.
+   */
+  recurring: boolean;
+
+  /**
+   * The cron expression or ISO 8601 duration string that defines the event's
+   * schedule.
+   */
+  schedule: string;
+};
+
+/**
+ * A cronjob caveat value that contains the jobs to be executed by the
+ * cronjob controller.
+ */
+type CronjobCaveatValue = {
+  expression: string;
+  request: {
+    method: string;
+    jsonrpc?: '2.0';
+    id?: string | number | null;
+    params?: Json[] | Record<string, Json>;
+  };
+};
+
+/**
+ * Get a duration with a minimum of 1 second. This function assumes the provided
+ * duration is valid.
+ *
+ * @param duration - The duration to validate.
+ * @returns The validated duration.
+ */
+// TODO: Export this function from the `@metamask/snaps-controllers` package?
+function getDuration(duration: Duration): Duration<true> {
+  if (duration.as('seconds') < 1) {
+    return Duration.fromObject({ seconds: 1 });
+  }
+
+  return duration;
+}
+
+/**
+ * Get the next execution date from a schedule, which should be either:
+ *
+ * - An ISO 8601 date string, or
+ * - An ISO 8601 duration string, or
+ * - A cron expression.
+ *
+ * @param schedule - The schedule of the event.
+ * @returns The parsed ISO 8601 date at which the event should be executed.
+ */
+// TODO: Export this function from the `@metamask/snaps-controllers` package?
+export function getExecutionDate(schedule: string) {
+  const date = DateTime.fromISO(schedule, { setZone: true });
+  if (date.isValid) {
+    const now = Date.now();
+
+    // We round to the nearest second to avoid milliseconds in the output.
+    const roundedDate = date.toUTC().startOf('second');
+    if (roundedDate.toMillis() < now) {
+      throw new Error('Cannot schedule an event in the past.');
+    }
+
+    return roundedDate.toISO({
+      suppressMilliseconds: true,
+    });
+  }
+
+  const duration = Duration.fromISO(schedule);
+  if (duration.isValid) {
+    // This ensures the duration is at least 1 second.
+    const validatedDuration = getDuration(duration);
+    return DateTime.now().toUTC().plus(validatedDuration).toISO();
+  }
```

Comment:
Is this case even possible before the migration?

Reply:
It's not, removed.